packages/babel-plugin-i18n-calypso/src/index.js

```diff
@@ -0,0 +1,202 @@
+
+/**
+ * Default output headers if none specified in plugin options.
+ *
+ * @type {Object}
+ */
+ const DEFAULT_HEADERS = {
+	'content-type': 'text/plain; charset=UTF-8',
+	'x-generator': 'babel-plugin-i18n-calypso',
+ };
+
+/**
+ * Given an argument node (or recursed node), attempts to return a string
+ * represenation of that node's value.
+ *
+ * @param {Object} node AST node.
+ *
+ * @return {string} String value.
+ */
+function getNodeAsString( node ) {
+	if ( undefined === node ) {
+		return '';
+	}
+
+	switch ( node.type ) {
+		case 'BinaryExpression':
+		return (
+			getNodeAsString( node.left ) +
+			getNodeAsString( node.right )
+			);
+
+		case 'StringLiteral':
+		return node.value;
+
+		default:
+		return '';
+	}
+}
+
+module.exports = function () {
+	let strings = {}, nplurals = 2, baseData;
+
+	return {
+		visitor: {
+			CallExpression( path, state ) {
+				const { callee } = path.node;
+
+				// Determine function name by direct invocation or property name
+				let name;
+				if ( 'MemberExpression' === callee.type ) {
+					name = callee.property.name;
+				} else {
+					name = callee.name;
+				}
+				if ( 'translate' !== name ) {
+					return;
+				}
+				let i = 0;
+
+				const translation = {
+					msgid: getNodeAsString( path.node.arguments[i++] ),
+					msgstr: '',
+					comments: {},
+				}
+
+				if ( ! translation.msgid.length ) {
+					return;
+				}
+
+				// At this point we assume we'll save data, so initialize if
+				// we haven't already
+				if ( ! baseData ) {
+					baseData = {
+						charset: 'utf-8',
+						headers: state.opts.headers || DEFAULT_HEADERS,
+						translations: {
+							'': {
+								'': {
+									msgid: '',
+									msgstr: [],
+								},
+							},
+						},
+					};
+
+					for ( const key in baseData.headers ) {
+						baseData.translations[ '' ][ '' ].msgstr.push( `${ key }: ${ baseData.headers[ key ] };\n` );
+					}
+
+					// Attempt to exract nplurals from header
+					const pluralsMatch = ( baseData.headers[ 'plural-forms' ] || '' ).match( /nplurals\s*=\s*(\d+);/ );
+					if ( pluralsMatch ) {
+						nplurals = pluralsMatch[ 1 ];
+					}
+				}
+
+				if ( path.node.arguments.length > i ) {
+					const msgid_plural = getNodeAsString( path.node.arguments[i] );
+					if ( msgid_plural.length ) {
+						translation.msgid_plural = msgid_plural;
+						i++;
+						// For plurals, create an empty mgstr array
+						translation.msgstr = Array.from( Array( nplurals ) ).map( () => '' );
+					}
+				}
+
+				const { filename } = this.file.opts;
+				const pathname = relative( '.', filename ).split( sep ).join( '/' );
+				translation.comments.reference = pathname + ':' + path.node.loc.start.line;
+
+				if ( path.node.arguments.length > i && 'ObjectExpression' === path.node.arguments[i].type ) {
+					for ( const j in path.node.arguments[i].properties ) {
+						if ( 'ObjectProperty' === path.node.arguments[i].properties[j].type ) {
```

Comment:
Maybe contract these to lines into 
```js
const args = path.node.arguments[i].properties.find( property => property.type === 'ObjectProperty' ) )
if ( args ) {
```

Reply:
Hm, how should this work since I am looping over all properties?

Reply:
Hmm, that's what the `find` does for you, doesn't it? Am I missing something?

This is what I mean:

```diff
diff --git a/packages/babel-plugin-i18n-calypso/src/index.js b/packages/babel-plugin-i18n-calypso/src/index.js
index 9968ae7f4c..d046f61e33 100644
--- a/packages/babel-plugin-i18n-calypso/src/index.js
+++ b/packages/babel-plugin-i18n-calypso/src/index.js
@@ -155,16 +155,15 @@ module.exports = function () {
                                translation.comments.reference = pathname + ':' + path.node.loc.start.line;
 
                                if ( path.node.arguments.length > i && 'ObjectExpression' === path.node.arguments[i].type ) {
-                                       for ( const j in path.node.arguments[i].properties ) {
-                                               if ( 'ObjectProperty' === path.node.arguments[i].properties[j].type ) {
-                                                       switch ( path.node.arguments[i].properties[j].key.name ) {
-                                                               case 'context':
-                                                                       translation.msgctxt = path.node.arguments[i].properties[j].value.value;
-                                                                       break;
-                                                               case 'comment':
-                                                                       translation.comments.extracted = path.node.arguments[i].properties[j].value.value;
-                                                                       break;
-                                                       }
+                                       const args = path.node.arguments[ i ].properties.find( property => property.type === 'ObjectProperty' );
+                                       if ( args ) {
+                                               switch ( args.key.name ) {
+                                                       case 'context':
+                                                               translation.msgctxt = args.value.value;
+                                                               break;
+                                                       case 'comment':
+                                                               translation.comments.extracted = args.value.value;
+                                                               break;
                                                }
                                        }
                                }
```

But granted, this is just syntactic sugar, so feel free to ignore.

Reply:
But a `translate()` call can have both `comment` and `context` entry, your approach doesn't cater for that, does it?

Reply:
Oh, okay. That's correct -- yours loops over all properties regardless, while mine returns upon first match. My bad, ignore then.